File: backend/providers/replicate.py
import httpx
from django.conf import settings
from .base import BaseProvider, SubmitInput, JobStatus

BASE = 'https://api.replicate.com/v1'


class ReplicateProvider(BaseProvider):
    name = 'replicate'

    def __init__(self, token: str | None = None):
        self._token = token or getattr(settings, 'REPLICATE_API_TOKEN', '')

    def _headers(self) -> dict:
        return {
            'Authorization': f'Bearer {self._token}',
            'Content-Type': 'application/json',
        }

    def submit(self, inp: SubmitInput) -> str:
        model_id = _model_id(inp.model_slug)
        payload: dict = {'input': inp.params}

        with httpx.Client(timeout=30) as client:
            if ':' in model_id:
                owner_name, version = model_id.rsplit(':', 1)
                r = client.post(f'{BASE}/predictions', headers=self._headers(), json={
                    'version': version,
                    'input': inp.params,
                })
            else:
                r = client.post(
                    f'{BASE}/models/{model_id}/predictions',
                    headers=self._headers(),
                    json=payload,
                )
            r.raise_for_status()
            return r.json()['id']

    def poll(self, provider_job_id: str, modality: str = 'image') -> JobStatus:
        with httpx.Client(timeout=15) as client:
            r = client.get(f'{BASE}/predictions/{provider_job_id}', headers=self._headers())
            r.raise_for_status()
            data = r.json()

        status_map = {
            'starting': 'processing',
            'processing': 'processing',
            'succeeded': 'completed',
            'failed': 'failed',
            'canceled': 'failed',
        }
        status = status_map.get(data['status'], 'processing')
        outputs: list[dict] = []

        if status == 'completed' and data.get('output'):
            raw = data['output']
            if isinstance(raw, str):
                raw = [raw]
            asset_type = modality if modality in ('image', 'video', 'audio') else 'image'
            outputs = [{'url': u, 'type': asset_type} for u in raw if isinstance(u, str)]

        return JobStatus(
            status=status,
            outputs=outputs,
            error=data.get('error'),
        )

    def cancel(self, provider_job_id: str) -> None:
        with httpx.Client(timeout=10) as client:
            client.post(f'{BASE}/predictions/{provider_job_id}/cancel', headers=self._headers())


def _model_id(model_slug: str) -> str:
    """Map registry slug → Replicate model id."""
    from providers.registry import MODEL_REGISTRY
    cfg = MODEL_REGISTRY.get(model_slug)
    if cfg and 'replicate' in cfg.get('providers', {}):
        return cfg['providers']['replicate']['model_id']
    return model_slug

File: frontend/src/lib/providers/replicate.ts
import type { GenerationProvider, ProviderJobStatus, SubmitInput } from './types';
import { modelRegistry } from '@/lib/models/registry';

const BASE = 'https://api.replicate.com/v1';

function headers() {
  const token = process.env.REPLICATE_API_TOKEN;
  if (!token) throw new Error('REPLICATE_API_TOKEN is not set');
  return {
    Authorization: `Bearer ${token}`,
    'Content-Type': 'application/json',
    Prefer: 'wait', // return immediately with a prediction id, don't block
  };
}

type ReplicatePrediction = {
  id: string;
  status: 'starting' | 'processing' | 'succeeded' | 'failed' | 'canceled';
  output: unknown;
  error: string | null;
  logs: string | null;
};

function mapStatus(p: ReplicatePrediction, modality: 'image' | 'video' | 'audio'): ProviderJobStatus {
  if (p.status === 'starting' || p.status === 'processing') {
    return { status: 'processing' };
  }
  if (p.status === 'succeeded') {
    const rawOutputs = Array.isArray(p.output) ? p.output : p.output ? [p.output] : [];
    return {
      status: 'completed',
      outputs: rawOutputs
        .filter((u): u is string => typeof u === 'string')
        .map((url) => ({ url, type: modality })),
    };
  }
  return {
    status: 'failed',
    error: p.error ?? p.status,
  };
}

export class ReplicateProvider implements GenerationProvider {
  readonly name = 'replicate' as const;

  async submit(input: SubmitInput): Promise<{ providerJobId: string }> {
    const model = modelRegistry[input.modelSlug];
    if (!model) throw new Error(`Unknown model: ${input.modelSlug}`);

    const cfg = model.providers.replicate;
    if (!cfg) throw new Error(`Model ${input.modelSlug} has no Replicate config`);

    const body: Record<string, unknown> = {
      input: input.params,
    };
    if (input.webhookUrl) {
      body.webhook = input.webhookUrl;
      body.webhook_events_filter = ['completed'];
    }

    // Replicate API: POST /models/{owner}/{name}/predictions  OR  /predictions with version
    const url = `${BASE}/models/${cfg.modelId}/predictions`;
    const res = await fetch(url, {
      method: 'POST',
      headers: headers(),
      body: JSON.stringify(body),
    });

    if (!res.ok) {
      const err = await res.text();
      throw new Error(`Replicate submit failed (${res.status}): ${err}`);
    }

    const data = (await res.json()) as ReplicatePrediction;
    return { providerJobId: data.id };
  }

  async poll(providerJobId: string): Promise<ProviderJobStatus> {
    const res = await fetch(`${BASE}/predictions/${providerJobId}`, {
      headers: headers(),
    });

    if (!res.ok) {
      throw new Error(`Replicate poll failed (${res.status})`);
    }

    const data = (await res.json()) as ReplicatePrediction;
    // We don't know modality here — caller must map outputs after the fact if needed
    return mapStatus(data, 'image'); // type will be overridden by caller from DB
  }

  async pollWithModality(
    providerJobId: string,
    modality: 'image' | 'video' | 'audio',
  ): Promise<ProviderJobStatus> {
    const res = await fetch(`${BASE}/predictions/${providerJobId}`, {
      headers: headers(),
    });

    if (!res.ok) {
      throw new Error(`Replicate poll failed (${res.status})`);
    }

    const data = (await res.json()) as ReplicatePrediction;
    return mapStatus(data, modality);
  }

  async cancel(providerJobId: string): Promise<void> {
    await fetch(`${BASE}/predictions/${providerJobId}/cancel`, {
      method: 'POST',
      headers: headers(),
    });
  }
}
